spec/integration/mentioning_spec.rb

```diff
@@ -19,77 +28,337 @@ def mention_stream_for(user)
-    expect(users_connected?(@user1, @user3)).to be false
+  RSpec::Matchers.define :be_in_streams_of do |user|
+    match do |status_message|
+      stream_for(user).map(&:id).include?(status_message.id) &&
+        mention_stream_for(user).map(&:id).include?(status_message.id)
+    end
 
+    match_when_negated do |status_message|
+      !stream_for(user).map(&:id).include?(status_message.id) &&
+        !mention_stream_for(user).map(&:id).include?(status_message.id)
+    end
+  end
+
+  let(:user1) { FactoryGirl.create(:user_with_aspect) }
+  let(:user2) { FactoryGirl.create(:user_with_aspect, friends: [user1, user3]) }
+  let(:user3) { FactoryGirl.create(:user_with_aspect) }
+
+  # see: https://github.com/diaspora/diaspora/issues/4160
+  it "only mentions people that are in the target aspect" do
     status_msg = nil
     expect {
-      post "/status_messages.json", status_message: {text: text_mentioning(@user3)}, aspect_ids: default_aspect.id.to_s
-      status_msg = StatusMessage.find(JSON.parse(response.body)["id"])
+      status_msg = post_status_message(user3)
     }.to change(Post, :count).by(1).and change(AspectVisibility, :count).by(1)
 
     expect(status_msg).not_to be_nil
     expect(status_msg.public?).to be false
-    expect(status_msg.text).to include(@user3.name)
-    expect(status_msg.text).not_to include(@user3.diaspora_handle)
-    expect(status_msg.text).to include(user_profile_path(username: @user3.username))
+    expect(status_msg.text).to include(user3.name)
 
-    expect(stream_for(@user3).map(&:id)).not_to include(status_msg.id)
-    expect(mention_stream_for(@user3).map(&:id)).not_to include(status_msg.id)
+    expect(user3).not_to be_mentioned_in(status_msg)
+    expect(status_msg).not_to be_in_streams_of(user3)
   end
 
-  it "mentions people in public posts" do
-    expect(users_connected?(@user1, @user3)).to be false
+  context "in private post via federation" do
+    let(:status_msg) {
+      post_status_message_via_federation(text_mentioning(user2) + text_mentioning(user3), [user3])
+    }
+
+    it "receiver is mentioned in status message" do
+      expect(user3).to be_mentioned_in(status_msg)
+    end
+
+    it "receiver can see status message in streams" do
+      expect(status_msg).to be_in_streams_of(user3)
+    end
+
+    it "non-receiver is not mentioned in status message" do
+      expect(user2).not_to be_mentioned_in(status_msg)
+    end
 
+    it "non-receiver can't see status message in streams" do
+      expect(status_msg).not_to be_in_streams_of(user2)
+    end
+  end
+
+  context "in private post via federation with multiple recipients" do
+    shared_context "in the first possible recipient order" do
+      let(:recipients) { [user3, user2] }
+    end
+
+    shared_context "in the second possible recipient order" do
+      let(:recipients) { [user2, user3] }
+    end
+
+    ["in the first possible recipient order", "in the second possible recipient order"].each do |in_context|
```

Comment:
But does the order matter then?

Maybe you can also add some more unit-tests for `Notifications::Mentioned.notify`?